### ISSUE TITLE
Translate Javascript SDK

### ISSUE THREAD

**Issue by mrozbarry** [pairshaped/elm-firebase#29]
# Preface

Firebase javascript sdk can be found [on their github page](https://github.com/firebase/firebase-js-sdk). This issue will be open to discuss general goals and hurdles in translating and maintaining the library.

# Goals/Discussion

We need to identify:

 - [ ] Required features that are currently **not** currently supported by Elm.
 - [ ] Elm libraries to minimize the amount of initial code we need to contribute.
 - [ ] Elm libraries we may have to produce as pre-requisites to translating firebase as a whole.
 - [ ] How to integrate new code from local issues/bugs.
 - [ ] How to integrate new code from upstream firebase updates.
 - [ ] Determine if there is anything in firebase.js that we **don't** actually need to support.

**Comment by ryanucode** [comment]
The firebase library implements a large number of compatibility shims and fallbacks, Notice here the realtime database functionality has [multiple implementations depending on the browser compatability](https://github.com/firebase/firebase-js-sdk/blob/master/src/database/js-client/realtime/TransportManager.js#L48). I (and my company) am interested in helping port the realtime database, auth, and storage functionality from the firebase library.

I read through a good chunk of their code base last weekend. The pain points of a port from what I  can see are:
1. They have a unit test around service workers. I have not found where they are used yet, but I do not know how well supported service workers are in Elm.
1. The code base is much larger if we intend to replicate a lot of their fallback / compatibility methods. Is this something you intend to support? (Like the long polling fallback for websockets)
1. The speed of upstream iteration is quick. (Firebase functions, phone auth, and updated initialization scheme in just the last month!)

Finally since now we can peak under the hood what are your thoughts on the shape of the public api?

**Comment by mrozbarry** [comment]
I can address some of what you've touched on:

## Unit tests around service workers

I don't think service workers are going to really be something we need to worry about from inside Elm. Service workers sit in a layer between developer javascript and the browser engine itself, mostly as a way to handle smarter caching or mimic online features while offline. Although an attractive feature of firebase is it's ability to go offline, I don't think it's a feature that we need to be concerned with until we have other features working. In fact, service workers can actually make debugging code and behaviours difficult since it can deal with browser-engine level caching.

## Fallback features

I think the initial push in elm-firebase should be to use the latest/best technologies. We can definitely look into fallbacks if/when they are necessary/requested. I think Elm itself has some issues running on IE9/10, so I think it's safe to initially target (in the IE/MS world) IE11/Edge. I think most other browsers play nice with Elm, and looks like everything (with IE11+, exception of Opera Mini) supports indexedDB, and websockets.

That said, I think we'll need to evaluate most of these fallbacks on a per-case basis. Some may be necessary to get all modern browsers on board.

## Upstream iteration

This is actually a big concern for me. It would be one thing if we had a magic `./js_to_elm firebase.js`, but since this is going to be translation by hand, and sometimes by nuance, it's going to be hard. I'm really interested and open to suggestions regarding a good process for triaging and applying upstream features.

Even more interesting is if/when we find bugs while translating. We'll need a process for documenting these bugs locally, and getting this documents posted upstream.

--

I'll have to get back to you on the public api, though, I haven't had a chance to do a thorough comb through.

**Comment by ryanucode** [comment]
I am not sure if this is the place to post technical details or if you want to collect this stuff in a wiki somewhere. I did more digging for fun tonight on what it would take to get the realtime database up and running. Couple of summary points.
1. Al teast a portion of the auth library would have to be implemented before the database functions (should have figured).
2. They use a versioned wire protocol / endpoints. I am not sure what the support lifetime of each version is, but this is good for us because it means we can continue communicating with older version endpoints when Google releases new ones.

What follows is the shortest i can manage "read me first" list. The files here are ordered by abstraction level roughly from low to high.

The wire protocol implementation:
https://github.com/firebase/firebase-js-sdk/blob/c054dab68ce4d30af7fcb0bb850f1a064ba023c5/src/database/third_party/closure-library/closure/goog/labs/net/webchannel/wirev8.js

The SDK wire to web channel internal interface. Potentially useful to follow their internal API's abstractions? It appears the wire protocol is a site of heavy development.:
https://github.com/firebase/firebase-js-sdk/blob/c054dab68ce4d30af7fcb0bb850f1a064ba023c5/src/database/third_party/closure-library/closure/goog/labs/net/webchannel/wire.js

Their "WebChannel implemenattion. This appears to manage connection state information, provides message queueing and retry.:
https://github.com/firebase/firebase-js-sdk/blob/c054dab68ce4d30af7fcb0bb850f1a064ba023c5/src/database/third_party/closure-library/closure/goog/labs/net/webchannel/webchannelbase.js

Web channel API. This appears to be an internal google library for websocket communications. Firebase contains a specialized internal fork according to the code comments. A web channel represents a logical connection pool abstracting web sockets / http. Their own explanation is here:
https://github.com/firebase/firebase-js-sdk/blob/c054dab68ce4d30af7fcb0bb850f1a064ba023c5/src/database/third_party/closure-library/closure/goog/labs/net/webchannel.js#L16

The Query implementation. This is the object they send to perform a database action:
https://github.com/firebase/firebase-js-sdk/blob/c054dab68ce4d30af7fcb0bb850f1a064ba023c5/src/database/js-client/core/view/QueryParams.js

**Comment by magicman86** [comment]
Is this still alive?

**Comment by mrozbarry** [comment]
Yes, just been busy with other things that have much higher priority than
this. PRs are welcome, of course.

**Comment by dstpierre** [comment]
@mrozbarry very interesting topic. I arrived here since I'm researching Elm and Firebase solutions.

Porting the JavaScript SDK appears to be a substantial task. I would be interested in contributing, is this GH issue still the right place to talk about this?

I'm not seeing a branch, how can I help?

Thanks
